src/transforms/gate/transform.rs

```diff
@@ -0,0 +1,268 @@
+use std::collections::VecDeque;
+
+use vector_lib::internal_event::{Count, InternalEventHandle as _, Registered};
+
+use crate::{
+    conditions::Condition,
+    event::Event,
+    internal_events::GateEventsDropped,
+    transforms::{FunctionTransform, OutputBuffer},
+    transforms::gate::state::GateState,
+};
+
+#[derive(Clone)]
+pub struct Gate {
+    // Configuration parameters
+    pass_when: Option<Condition>,
+    open_when: Option<Condition>,
+    close_when: Option<Condition>,
+    max_events: usize,
+    auto_close: bool,
+    tail_events: usize,
+
+    // Internal variables
+    current_state: GateState,
+    buffer: VecDeque<Event>,
+    events_counter: usize,
+    events_dropped: Registered<GateEventsDropped>,
+    is_closing: bool,
+}
+
+impl Gate {
+    // This function is dead code when the feature flag `transforms-impl-gate` is specified but not
+    // `transforms-gate`.
+    #![allow(dead_code)]
+    pub fn new(
+        pass_when: Option<Condition>,
+        open_when: Option<Condition>,
+        close_when: Option<Condition>,
+        max_events: usize,
+        auto_close: bool,
+        tail_events: usize,
+    ) -> crate::Result<Self> {
+        let buffer = VecDeque::with_capacity(max_events);
+
+        Ok(Gate {
+            pass_when,
+            open_when,
+            close_when,
+            max_events,
+            auto_close,
+            tail_events,
+            events_dropped: register!(GateEventsDropped),
+            current_state: GateState::Closed,
+            buffer,
+            events_counter: 0,
+            is_closing: false,
+        })
+    }
+}
+
+impl FunctionTransform for Gate {
+    fn transform(&mut self, output: &mut OutputBuffer, event: Event) {
+        let (pass_gate, event) = match self.pass_when.as_ref() {
```

Comment:
I think you could return early after this block if `pass_gate` is true to avoid evaluating the other conditions and pushing/popping from the dequeue.